cpp/src/reductions/segmented/valid_counts.cu

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "valid_counts.hpp"
+
+#include <cudf/detail/null_mask.cuh>
+
+#include <thrust/adjacent_difference.h>
+
+namespace cudf {
+namespace reduction {
+namespace detail {
+
+rmm::device_uvector<size_type> segmented_valid_counts(column_device_view const& d_col,
+                                                      bool has_nulls,
+                                                      device_span<size_type const> offsets,
+                                                      null_policy null_handling,
+                                                      rmm::cuda_stream_view stream,
+                                                      rmm::mr::device_memory_resource* mr)
+{
+  auto const num_segments = offsets.size() - 1;
+
+  if (has_nulls && (null_handling == null_policy::EXCLUDE)) {
+    return cudf::detail::segmented_count_bits(d_col.null_mask(),
```

Comment:
If we only use `d_col` for the null mask here, why do we bother to build the device view?

Reply:
Good point. We only really need to pass the null-mask.